From b1e710134271e3e537d89061437aa1894aedfd72 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Fri, 5 Sep 2014 01:33:36 -0400
Subject: [iommu] vt-d: fix wrong return value of dmar_table_init()

Message-id: <20140905013336.30867.98193.stgit@gir.stowe>
Patchwork-id: 91289
O-Subject: [RHEL7.1 PATCH 25/28] iommu/vt-d: fix wrong return value of dmar_table_init()
Bugzilla: 1129808
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Dutile <ddutile@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1129808
Upstream Status: cc05301fd54f3e166aedf24e39f6731c4dec0451

Signed-off-by: Jarod Wilson <jarod@redhat.com>

Conflicts:
  ./drivers/iommu/dmar.c
    Hunk #1 suffex context - Due to RHEL7 commit 12c00b18dd8.

commit cc05301fd54f3e166aedf24e39f6731c4dec0451
Author: Jiang Liu <jiang.liu@linux.intel.com>
Date:   Mon Jan 6 14:18:24 2014 +0800

    iommu/vt-d: fix wrong return value of dmar_table_init()

    If dmar_table_init() fails to detect DMAR table on the first call,
    it will return wrong result on following calls because it always
    sets dmar_table_initialized no matter if succeeds or fails to
    detect DMAR table.

    Signed-off-by: Jiang Liu <jiang.liu@linux.intel.com>
    Signed-off-by: Joerg Roedel <joro@8bytes.org>
---

diff --git a/drivers/iommu/dmar.c b/drivers/iommu/dmar.c
index 67473a8..808d2dd 100644
--- a/drivers/iommu/dmar.c
+++ b/drivers/iommu/dmar.c
@@ -472,24 +472,23 @@ int __init dmar_table_init(void)
  static int dmar_table_initialized;
  int ret;
 
- if (dmar_table_initialized)
-  return 0;
-
- dmar_table_initialized = 1;
-
- ret = parse_dmar_table();
- if (ret) {
-  if (ret != -ENODEV)
-   pr_info("parse DMAR table failure.\n");
-  return ret;
- }
+ if (dmar_table_initialized == 0) {
+  ret = parse_dmar_table();
+  if (ret < 0) {
+   if (ret != -ENODEV)
+    pr_info("parse DMAR table failure.\n");
+  } else  if (list_empty(&dmar_drhd_units)) {
+   pr_info("No DMAR devices found\n");
+   ret = -ENODEV;
+  }
 
- if (list_empty(&dmar_drhd_units)) {
-  pr_info("No DMAR devices found\n");
-  return -ENODEV;
+  if (ret < 0)
+   dmar_table_initialized = ret;
+  else
+   dmar_table_initialized = 1;
  }
 
- return 0;
+ return dmar_table_initialized < 0 ? dmar_table_initialized : 0;
 }
 
 static inline void warn_invalid_dmar(u64 addr, const char *message)
-- 
1.7.1